### PR TITLE
1985 parser owl functional syntax

### PR DESCRIPTION
**Pull request by b-gehrke**
Implemented parser for OWL functional syntax to abstract syntax

### CODE REVIEW

**Review by mcodescu — COMMENTED** [review]
I have tested the parser on a couple of examples and it seems to be working. Before I do more tests, there are a couple of things that we should discuss:
1. the old parser for Manchester syntax took an argument `GA.prefixMap` such that the prefixes declared in a .dol file are visible. @tillmo do we have a reason not to do this anymore?
2.  currently the entities appear in the sentences in their abbreviated form, e.g. `Assertion (ObjectPropertyAssertion [] (ObjectProp :hasMother) (NamedIndividual_ :Stewie) (NamedIndividual_ :Lois))`. Thus we won't be able to fix #1563 even after switching to OWL functional syntax. It seems to me we should expand prefixes and the pretty printer should display the entities using the prefix if the prefix is still available (the static analysis may remove duplicated conflicting prefixes). @tillmo do we want to do this already?

File: OWL2/ParseAS.hs [hunk 1]
```diff
@@ -0,0 +1,754 @@
+{-# LANGUAGE TupleSections #-}
+
+module OWL2.ParserAS where
```
[the remark this answers is not in the document]
Comment (b-gehrke) [reply OWL2/ParseAS.hs]: Renamed module, as the pattern for parser file names seems to be `Parse[lang].hs`

**Comment by b-gehrke** [comment]
The parser now accepts a PrefixMap and can now expand IRIs while parsing.

**Comment by mcodescu** [comment]
I have added a tracing message after the expansion of the IRIs, and I used the method `showTrace` from `Common.IRI`, extended to show the `iFragment` as well, because it shows us the values of each field of the record. With `test3.ofn` from the tests folder in OWL, I get for example:
```
scheme:
no authority
path::hasChild
query:
fragment:
prefix:
iFragment:
isAbbrev:True
```
This is an abbreviated IRI `:hasChild` with the prefix `: = <http://example.org/>`, and it should be expanded. Now, the problem seems to be the handling of the empty prefix. In `expandIRI` you do `lookup (prefixName iri) pm`, but the prefix map is `pm:fromList [(":",<http://example.org/>)]` and the prefix is the empty string, so the expansion does not take place.

**Comment by b-gehrke** [comment]
> This is an abbreviated IRI `:hasChild` with the prefix `: = <http://example.org/>`, and it should be expanded. Now, the problem seems to be the handling of the empty prefix. In `expandIRI` you do `lookup (prefixName iri) pm`, but the prefix map is `pm:fromList [(":",<http://example.org/>)]` and the prefix is the empty string, so the expansion does not take place.

The Manchester parser doesn't add the colon to the prefixes. So I changed the AS parser accordingly

**Comment by b-gehrke** [comment]
> ```
> scheme:
> no authority
> path::hasChild
> query:
> fragment:
> prefix:
> iFragment:
> isAbbrev:True
> ```

Also the `iFragment` shouldn't be empty and the `path` shouldn't contain the colon. I looked into it and the curie parser doesn't accept the empty prefix. With the colon being part of the `path` the expanded IRI would look like this `<http://example.org/:hasChild>`. I'm not sure if this issue should be addressed in this PR.

**Comment by b-gehrke** [comment]
With the extended IRI parser all bioportal ontologies can be parsed except the following due to the use of `DLSafeRule`:
- ADAR
- BOF
- BRIDG
- FHHO
- OF
- OGDI
- SITBAC
- STATO

**Pull request merged**